// sqreen/test/lib/instrumentation/hooks/knex.js
/**
 * Copyright (c) 2016 - 2018 Sqreen. All Rights Reserved.
 * Please refer to our terms for more information: https://www.sqreen.io/terms.html
 */
'use strict';
const Code = require('code');
const Lab = require('lab');
const lab = exports.lab = Lab.script();

const describe = lab.describe;
const it = lab.it;
const expect = Code.expect;

const Hook = require('../../../../lib/instrumentation/hooks/knexHook');
const NS = require('../../../../lib/instrumentation/hooks/ns').getNS();

describe('knex hook', () => {

    it('should test the hook', { plan: 2 }, (done) => {

        let i = 0;
        const Client = class {

            queryBuilder() {

                return {
                    then: function (cb) {

                        return cb();
                    }
                };
            }

            runner() {

                return {
                    client: {
                        releaseConnection: function () {

                            i++;
                            if (i === 2) {
                                done();
                            }
                        }
                    }
                };
            }
        };

        const mod = {
            Client
        };

        Hook({}, mod);

        const client = new Client();
        client.runner();
        NS.run(() => {

            NS.set('a', 'hello');

            client.queryBuilder()
                .then(() => {

                    expect(NS.get('a')).to.equal('hello');
                });

            client.runner();
            expect(NS.get('raise')).to.have.length(1);

            NS.get('raise')[0]();
            client.runner();
            NS.get('raise')[1]();

        });
    });
});
